AsyncDisplayKit/TextKit/ASTextKitFontSizeAdjuster.mm

```diff
@@ -0,0 +1,185 @@
+{
+  __weak ASTextKitContext *_context;
+  ASTextKitAttributes _attributes;
+  std::mutex _textKitMutex;
+}
+
+- (instancetype)initWithContext:(ASTextKitContext *)context
+                constrainedSize:(CGSize)constrainedSize
+              textKitAttributes:(const ASTextKitAttributes &)textComponentAttributes;
+{
+  if (self = [super init]) {
+    _context = context;
+    _constrainedSize = constrainedSize;
+    _attributes = textComponentAttributes;
+  }
+  return self;
+}
+
++ (void)adjustFontSizeForAttributeString:(NSMutableAttributedString *)attrString withScaleFactor:(CGFloat)scaleFactor
+{
+  [attrString beginEditing];
+
+  // scale all the attributes that will change the bounding box
+  [attrString enumerateAttributesInRange:NSMakeRange(0, attrString.length) options:0 usingBlock:^(NSDictionary<NSString *,id> * _Nonnull attrs, NSRange range, BOOL * _Nonnull stop) {
+    if (attrs[NSFontAttributeName] != nil) {
+      UIFont *font = attrs[NSFontAttributeName];
+      font = [font fontWithSize:roundf(font.pointSize * scaleFactor)];
+      [attrString removeAttribute:NSFontAttributeName range:range];
+      [attrString addAttribute:NSFontAttributeName value:font range:range];
+    }
+    
+    if (attrs[NSKernAttributeName] != nil) {
+      NSNumber *kerning = attrs[NSKernAttributeName];
+      [attrString removeAttribute:NSKernAttributeName range:range];
+      [attrString addAttribute:NSKernAttributeName value:@([kerning floatValue] * scaleFactor) range:range];
+    }
+    
+    if (attrs[NSParagraphStyleAttributeName] != nil) {
+      NSMutableParagraphStyle *paragraphStyle = [attrs[NSParagraphStyleAttributeName] mutableCopy];
+      paragraphStyle.lineSpacing = (paragraphStyle.lineSpacing * scaleFactor);
+      paragraphStyle.paragraphSpacing = (paragraphStyle.paragraphSpacing * scaleFactor);
+      paragraphStyle.firstLineHeadIndent = (paragraphStyle.firstLineHeadIndent * scaleFactor);
+      paragraphStyle.headIndent = (paragraphStyle.headIndent * scaleFactor);
+      paragraphStyle.tailIndent = (paragraphStyle.tailIndent * scaleFactor);
+      paragraphStyle.minimumLineHeight = (paragraphStyle.minimumLineHeight * scaleFactor);
+      paragraphStyle.maximumLineHeight = (paragraphStyle.maximumLineHeight * scaleFactor);
+      paragraphStyle.lineHeightMultiple = (paragraphStyle.lineHeightMultiple * scaleFactor);
+      paragraphStyle.paragraphSpacing = (paragraphStyle.paragraphSpacing * scaleFactor);
+      
+      [attrString removeAttribute:NSParagraphStyleAttributeName range:range];
+      [attrString addAttribute:NSParagraphStyleAttributeName value:paragraphStyle range:range];
+    }
+    
+  }];
+
+  [attrString endEditing];
+}
+
+- (NSUInteger)lineCountForString:(NSAttributedString *)attributedString
+{
+  NSUInteger lineCount = 0;
+  
+  static std::mutex __static_mutex;
+  std::lock_guard<std::mutex> l(__static_mutex);
+  
+  NSTextStorage *textStorage = [[NSTextStorage alloc] initWithAttributedString:attributedString];
```

Comment:
Using the current storage, manager and container didn't seem to work, so I made new ones. Not sure if I need the lock here since we are already in the _context's lock.